File: app/src/main/java/rbsoftware/friendstagram/ui/adapter/ProfileAdapter.java
package rbsoftware.friendstagram.ui.adapter;

import android.databinding.DataBindingUtil;
import android.net.Uri;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import java.util.ArrayList;
import java.util.List;

import rbsoftware.friendstagram.ActionExecuteListener;
import rbsoftware.friendstagram.Constants;
import rbsoftware.friendstagram.ui.viewholder.PictureViewHolder;
import rbsoftware.friendstagram.R;
import rbsoftware.friendstagram.databinding.HeaderProfileBinding;
import rbsoftware.friendstagram.model.Post;
import rbsoftware.friendstagram.model.User;

/**
 * Created by silver_android on 13/10/16.
 */

public class ProfileAdapter extends RecyclerView.Adapter {

    private static final int ITEM_VIEW_TYPE_HEADER = 0;
    private static final int ITEM_VIEW_TYPE_ITEM = 1;
    private User user;
    private ArrayList<Post> posts;
    private final ImageClickListener listener;
    private final ActionExecuteListener actionExecuteListener;

    public ProfileAdapter(ImageClickListener imageClickListener, ActionExecuteListener actionExecuteListener) {
        listener = imageClickListener;
        this.actionExecuteListener = actionExecuteListener;
        posts = new ArrayList<>();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == ITEM_VIEW_TYPE_HEADER) {
            HeaderProfileBinding viewDataBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
                    R.layout.header_profile, parent, false);
            return new HeaderViewHolder(viewDataBinding, actionExecuteListener);
        }

        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_image, parent, false);
        return new PictureViewHolder(view);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder parent, int position) {
        if (isHeader(position)) {
            HeaderViewHolder holder = (HeaderViewHolder) parent;
            holder.itemView.setUser(user);
        } else {
            final PictureViewHolder holder = (PictureViewHolder) parent;
            int index = position - (user == null ? 0 : 1);
            final Post post = posts.get(index);
            final boolean[] imageLoaded = {false};
            holder.image.setImageURI(Uri.parse(post.getImageURL()));
            imageLoaded[0] = true;
            holder.image.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (imageLoaded[0]) {
                        listener.onImageClick(post);
                    }
                }
            });
        }
    }

    public void setPosts(ArrayList<Post> posts) {
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new PostsDiffCallback(this.posts, posts));
        this.posts.clear();
        this.posts.addAll(posts);
        diffResult.dispatchUpdatesTo(this);
    }

    @Override
    public int getItemCount() {
        return posts.size() + (user != null ? 1 : 0);
    }

    @Override
    public int getItemViewType(int position) {
        return isHeader(position) ? ITEM_VIEW_TYPE_HEADER : ITEM_VIEW_TYPE_ITEM;
    }

    public void setUser(User user) {
        boolean hadNoUser = this.user == null;
        this.user = user;

        if (hadNoUser)
            notifyItemInserted(0);
        else
            notifyItemChanged(0);

        setPosts(user.getPosts());
    }

    public boolean isHeader(int position) {
        return user != null && position == 0;
    }

    public class HeaderViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private static final String TAG = "HeaderViewHolder";
        private final HeaderProfileBinding itemView;
        private final ActionExecuteListener listener;
        private final ImageView editProfile;

        HeaderViewHolder(HeaderProfileBinding itemView, ActionExecuteListener listener) {
            super(itemView.getRoot());
            this.itemView = itemView;
            this.listener = listener;
            this.itemView.executePendingBindings();
            editProfile = (ImageView) itemView.getRoot().findViewById(R.id.edit_profile);
            editProfile.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            Log.d(TAG, "editProfileClick: Edit Profile");
            listener.runAction(Constants.Action.EDIT_PROFILE);
        }
    }

    public interface ImageClickListener {
        void onImageClick(Post post);
    }

    private class PostsDiffCallback extends DiffUtil.Callback {

        private final List<Post> oldPosts;
        private final List<Post> newPosts;

        PostsDiffCallback(List<Post> oldPosts, List<Post> newPosts) {
            this.oldPosts = oldPosts;
            this.newPosts = newPosts;
        }

        @Override
        public int getOldListSize() {
            return oldPosts.size();
        }

        @Override
        public int getNewListSize() {
            return newPosts.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return oldPosts.get(oldItemPosition).equals(newPosts.get(newItemPosition));
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            return oldPosts.get(oldItemPosition).equals(newPosts.get(newItemPosition));
        }
    }
}

File: app/src/main/java/rbsoftware/friendstagram/ui/adapter/ProfileAdapter.kt
package rbsoftware.friendstagram.ui.adapter

import android.net.Uri
import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import io.reactivex.subjects.PublishSubject
import rbsoftware.friendstagram.Actions
import rbsoftware.friendstagram.GenericDiffCallback
import rbsoftware.friendstagram.R
import rbsoftware.friendstagram.inflate
import rbsoftware.friendstagram.model.Action
import rbsoftware.friendstagram.model.Post
import rbsoftware.friendstagram.model.User
import rbsoftware.friendstagram.ui.viewholder.PictureViewHolder

/**
 * Created by Rushil on 8/18/2017.
 */
class ProfileAdapter(private var posts: List<Post>, private var user: User, private var isMe: Boolean, private var isFollowing: Boolean) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private val onPostSelected: PublishSubject<Post> = PublishSubject.create()
    private val onActionExecuted: PublishSubject<Action> = PublishSubject.create()

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == ITEM_VIEW_TYPE_HEADER) {
            val view = parent?.inflate(R.layout.header_profile)
            HeaderViewHolder(view, isMe, isFollowing)
        } else {
            val view = parent?.inflate(R.layout.item_image)
            PictureViewHolder(view)
        }
    }

    override fun onBindViewHolder(parent: RecyclerView.ViewHolder?, position: Int) {
        if (isHeader(position)) {
            val holder = parent as HeaderViewHolder
            with(user) {
                holder.name?.text = name
                holder.username?.text = username
                holder.description?.text = description
            }
        } else {
            val holder = parent as PictureViewHolder
            val post = posts[position - 1]
            holder.image?.setImageURI(Uri.parse(post.imageURL))
            holder.image?.setOnClickListener { onPostSelected.onNext(post) }
        }
    }

    override fun getItemCount(): Int {
        return posts.size + 1
    }

    override fun getItemViewType(position: Int): Int {
        return if (isHeader(position)) {
            ITEM_VIEW_TYPE_HEADER
        } else {
            ITEM_VIEW_TYPE_ITEM
        }
    }

    fun setUser(user: User, isMe: Boolean, isFollowing: Boolean) {
        this.user = user
        this.isMe = isMe
        this.isFollowing = isFollowing
        notifyItemChanged(0)
        setPosts(user.posts)
    }

    fun isHeader(position: Int): Boolean {
        return position == 0
    }

    fun getOnPostSelected(): PublishSubject<Post> = onPostSelected

    fun getOnActionExecuted(): PublishSubject<Action> = onActionExecuted

    private fun setPosts(posts: List<Post>) {
        val diffResult = DiffUtil.calculateDiff(GenericDiffCallback(this.posts, posts))
        this.posts = List(posts.size, { i -> posts[i] })
        diffResult.dispatchUpdatesTo(this)
    }

    companion object {
        private const val ITEM_VIEW_TYPE_HEADER = 0
        private const val ITEM_VIEW_TYPE_ITEM = 1
    }

    inner class HeaderViewHolder internal constructor(containerView: View?, isMe: Boolean, isFollowing: Boolean) : RecyclerView.ViewHolder(containerView) {
        private val editProfile: ImageView? = containerView?.findViewById(R.id.edit_profile)
        private val followUser: ImageView? = containerView?.findViewById(R.id.follow_user)
        private val unfollowUser: ImageView? = containerView?.findViewById(R.id.unfollow_user)
        val name: TextView? = containerView?.findViewById(R.id.name)
        val username: TextView? = containerView?.findViewById(R.id.username)
        val description: TextView? = containerView?.findViewById(R.id.description)

        init {
            if (isMe) {
                editProfile?.visibility = VISIBLE
                followUser?.visibility = GONE
                unfollowUser?.visibility = GONE
                editProfile?.setOnClickListener {
                    onActionExecuted.onNext(Action(
                            Actions.EDIT_PROFILE,
                            mapOf(
                                    "user" to user
                            )
                    ))
                }
            } else {
                if (isFollowing) {
                    unfollowUser?.visibility = VISIBLE
                    editProfile?.visibility = GONE
                    followUser?.visibility = GONE
                    unfollowUser?.setOnClickListener {
                        onActionExecuted.onNext(Action(
                                Actions.UNFOLLOW_USER,
                                emptyMap()
                        ))
                    }
                } else {
                    followUser?.visibility = VISIBLE
                    editProfile?.visibility = GONE
                    unfollowUser?.visibility = GONE
                    followUser?.setOnClickListener {
                        onActionExecuted.onNext(Action(
                                Actions.FOLLOW_USER,
                                emptyMap()
                        ))
                    }
                }
            }
        }
    }
}